### DISCUSSION TITLE
env vars vs. bootstrapConfig

### DISCUSSION THREAD

**Discussion by klemen-df** [emqx/emqx-operator#700] [Q&A] [answered]
Hi,

I've read that _bootstrapConfig is only allowed to be configured when creating an EMQX cluster, and does not support updating. The bootstrap config can only be set at creation time or modified via the dashboard._

Also, found this : 'Configuration priority: emqx.conf < ENV < cluster-overrride.conf' - not really well documented, btw :)

Anyhow, what about scenario **where I don't have any bootstrapConfig**? Where all of the configuration is an environment variable?

What happens if I create a cluster (for example) with: 
EMQX_LISTENER_SSL_OPTS__VERSIONS=["tlsv1.3","tlsv1.2","tlsv1.1"]

and then I change the CRD and (try to) apply it with:
EMQX_LISTENER_SSL_OPTS__VERSIONS=["tlsv1.3","tlsv1.2"]

(variable being - listener_ssl_opts.versions, [here](https://www.emqx.io/docs/en/v5.0/configuration/configuration-manual.html#others))

Will it (should it) be applied?

**Comment by Rory-Z** [comment]
Hi, @KlemenDanfoss
> Anyhow, what about scenario where I don't have any bootstrapConfig? Where all of the configuration is an environment variable?

If you don't set any bootstrapConfig, the EMQX Operator will add some default config for EMQX custom resource, you can check it by `kubectl get emqx emqx -o json | jq .spec.bootstrapConfig` after deploy EMQX.

>What happens if I create a cluster (for example) with:
>EMQX_LISTENER_SSL_OPTS__VERSIONS=["tlsv1.3","tlsv1.2","tlsv1.1"]
>and then I change the CRD and (try to) apply it with:
>EMQX_LISTENER_SSL_OPTS__VERSIONS=["tlsv1.3","tlsv1.2"]

EMQX Operator will not enable SSL listener by default, so I'm not sure what happened, @zhongwencool could you please take a look ?

**Reply by zhongwencool** [reply]
It's useless for you to `EMQX_LISTENER_SSL_OPTS__VERSIONS=["tlsv1.3","tlsv1.2"]`. 
If you don't want to change via bootstrapConfig, 
You have to write the entire SSL listener configuration to the environment variable.
`EMQX_LISTENER_SSL_DEFAULT={
  bind = "0.0.0.0:8883"
  max_connections = 512000
  ssl_options {
    keyfile = "etc/certs/key.pem"
    certfile = "etc/certs/cert.pem"
    cacertfile = "etc/certs/cacert.pem"
  }
}`

**Comment by klemen-df** [comment, marked as answer]
So, scenario where it works:

- nothing set (empty) in bootstrapConfig
- EMQX_LISTENERS__SSL__MQTTS__BIND="0.0.0.0:8883"
- EMQX_LISTENERS__SSL__MQTTS__SSL_OPTIONS__VERSIONS=["tlsv1.3","tlsv1.2"]
- EMQX_LISTENERS__SSL__MQTTS__SSL_OPTIONS__KEYFILE="/folder/keyfile.pem"
- EMQX_LISTENERS__SSL__MQTTS__SSL_OPTIONS__CERTFILE="/folder/certfile.pem"
- EMQX_LISTENERS__SSL__MQTTS__SSL_OPTIONS__CACERTFILE="/folder/cacertfile.pem"

**Reply by Rory-Z** [reply]
Yes, also you can set `bootstrapConfig` like this:
```
apiVersion: apps.emqx.io/v2alpha1
kind: EMQX
metadata:
  name: emqx
spec:
  image: "emqx:5.0"
  bootstrapConfig: |
    listeners.ssl.mqtts {
      bind = "0.0.0.0:8883"
      ssl_options {
        versions = ["tlsv1.3","tlsv1.2"]
        keyfile = "/folder/keyfile.pem"
        certfile = "/folder/certfile.pem"
        cacertfile = "/folder/certfile.pem"
      }
    }
```
For EMQX Broker, they are the same

**Reply by klemen-df** [reply]
Yes, but the point is ; can (or can't) change bootstrapConfig after the initial run (but you can env vars)?

Or, at least, how it's supposed to work?

**Reply by Rory-Z** [reply]
Yes, the bootstrapConfig can not change after the initial run, but environment vars can be changed.

But please be careful, the modified environment variables may not take effect.

In EMQX application, the configuration priority is `etc/emqx.conf (bootstrapConfig) < ENV < data/cluster-overrride.conf`, the changes to the configuration via Dashboard will go into `data/cluster-override.conf`, this means that if you modify the listener via EMQX Dashboard, the environment variables will not take effect even if you update them.

So if you want change EMQX config after initial run, the bast way is via EMQX Dashboard. (that's way the bootstrapConfig can not change after initial run.)

**Reply by klemen-df** [reply]
Hi,

thanks for the clarification. Probably this should be documented better.

But in our use-case - we use Terraform/IaaC - we don't make any changes through the dashboard so env vars are fine.

**Reply by Rory-Z** [reply]
Very cool !

**Reply by alexa22013** [reply]
Hey, @Rory-Z as you mentioned
> Yes, the bootstrapConfig can not change after the initial run, but environment vars can be changed.

But as per the github documentation https://github.com/emqx/emqx-operator/blob/main/docs/en_US/reference/v2alpha1-reference.md#:~:text=List%20of%20environment%20variables%20to%20set%20in%20the%20container.%20Cannot%20be%20updated envs cannot be updated

So is it possible to update config dynamically or atleast are a rolling update in gke cluster, and also without using the dashboard?

**Reply by klemen-df** [reply]
I changed tls settings (just the env var) and it worked.

**Reply by Rory-Z** [reply]
@alexa22013 

> But as per the github documentation https://github.com/emqx/emqx-operator/blob/main/docs/en_US/reference/v2alpha1-reference.md#:~:text=List%20of%20environment%20variables%20to%20set%20in%20the%20container.%20Cannot%20be%20updated envs cannot be updated

This "cannot be updated" means if update env vars, the container will restart. So you can update env vars, and wait EMQX containers rolling updated
